### PR TITLE
Split subclasses out of ArchetypeType.

### PR DESCRIPTION
**Pull request by jckarter**
Context archetypes and opened existential archetypes differ in a number of details, and this simplifies the overlapping storage of the kind-specific fields. This should be NFC; for now, this doesn't change the interface of ArchetypeType, but should allow some refinements of how the special handling of certain archetypes are handled.

### CODE REVIEW

File: include/swift/AST/TypeNodes.def [hunk 1]
```diff
@@ -131,9 +131,13 @@ ABSTRACT_TYPE(AnyMetatype, Type)
 ALWAYS_CANONICAL_TYPE(Module, Type)
 TYPE(DynamicSelf, Type)
 ABSTRACT_TYPE(Substitutable, Type)
-  ALWAYS_CANONICAL_TYPE(Archetype, SubstitutableType)
+  ABSTRACT_TYPE(Archetype, SubstitutableType)
+    ALWAYS_CANONICAL_TYPE(ContextArchetype, ArchetypeType)
```
Comment (slavapestov) [inline include/swift/AST/TypeNodes.def]: I hate to make you rename everything but we've called these "primary archetypes" elsewhere

**Review by slavapestov — COMMENTED** [review]
Don't forget to update LLDB which annoyingly switches over TypeKind a lot

File: include/swift/AST/TypeNodes.def (hunk 1 above)
Comment (jckarter) [reply include/swift/AST/TypeNodes.def]: No problem. I'll also rename `ChildArchetype` to `NestedArchetype`, which jives better with the terminology used elsewhere.

**Comment by jckarter** [comment]
Please test with following pull request:
https://github.com/apple/swift-lldb/pull/1133

@swift-ci Please test

**Comment by jckarter** [comment]
@slavapestov How's this look now?

**Pull request merged**